File: Core/src/main/java/com/inverce/mod/core/collections/CollectionsEx.java
package com.inverce.mod.core.collections;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

public class CollectionsEx {
    @NonNull
    public static <T> List<T> join(@Nullable List<? extends T> A, @Nullable List<? extends T> B) {
        ArrayList<T> list = A != null ? new ArrayList<T>(A) : new ArrayList<T>();
        if (B != null) {
            list.addAll(B);
        }
        return list;
    }

    public static <T> boolean equals(@Nullable List<? extends T> A, @Nullable List<? extends T> B) {
        if (A == null || B == null) return A == B;
        if (A.size() != B.size()) return false;

        for (int i = 0; i < A.size(); i++) {
            if (!A.get(i).equals(B.get(i))) {
                return false;
            }
        }
        return true;
    }
}

File: Core/src/main/java/com/inverce/mod/v2/core/collection/CollectionsEx.kt
@file:JvmName("CollectionsEx")

package com.inverce.mod.v2.core.collection

fun <T> (List<T>?).sameAs(B: List<T>?): Boolean {
    if (this == null || B == null) return this === B
    if (this.size != B.size) return false

    for (i in this.indices) {
        if (this[i] != B[i]) {
            return false
        }
    }
    return true
}

fun <T : Y, Y> List<T>.migrateType() : List<T> = this.map { it }